server/broadcast.go

```diff
@@ -473,34 +475,55 @@ func processSegment(cxn *rtmpConnection, seg *stream.HLSSegment) ([]string, erro
 		sv = verification.NewSegmentVerifier(Policy)
 	}
 
-	for i := 0; i < MaxAttempts; i++ {
-		// if fails, retry; rudimentary
-		var urls []string
-		if urls, err = transcodeSegment(cxn, seg, name, sv); err == nil {
-			return urls, nil
+	var (
+		startTime = time.Now()
+		attempts  []data.TranscodeAttemptInfo
+		urls      []string
+	)
+	for len(attempts) < MaxAttempts {
+		// if transcodeSegment fails, retry; rudimentary
+		var info data.TranscodeAttemptInfo
+		urls, info, err = transcodeSegment(cxn, seg, name, sv)
+		attempts = append(attempts, info)
+		if err == nil {
+			break
 		}
 
 		if shouldStopStream(err) {
 			glog.Warningf("Stopping current stream due to err=%v", err)
 			rtmpStrm.Close()
-			return nil, err
+			break
 		}
-
 		if isNonRetryableError(err) {
 			glog.Warningf("Not retrying current segment nonce=%d seqNo=%d due to non-retryable error err=%v", nonce, seg.SeqNo, err)
-			return nil, err
+			break
 		}
-
 		// recoverable error, retry
 	}
-	if err != nil {
+	if MetadataQueue != nil {
+		var (
+			key      = fmt.Sprintf("stream_health.transcode.%c.%s", mid[0], mid)
```

Comment:
Why are using the first character of the manifestID followed by the full manifestID?

Reply:
I'll make this clearer by creating a temp var before this one, specifying what it means.

Context:
We use the first character in the manifest ID as a sharding key between some of the infrastructure components, like the broadcaster specifically. So this first character is like a "shard key" that could be used by consumers to only get the events from a specific shard. That would be necessary if we also want to shard the stream health (`data/analyzer`) nodes that consume these events.

By default, which is how we are going to start, the `analyzer` nodes just subscribe to `*` on that routing key segment so there's no sharding. We already have all the paraphernalia in place to turn it on though, in which case an `analyzer` from shard like `abcd` (all streams starting with a, b, c or d) will subscribe (bind their rabbitmq stream) to the topics having each of those letters (low level: they actually create N bindings, one for each letter). This is the code snippet that does that: [analyzer/reducers/transcode.go](https://github.com/livepeer/livepeer-data/blob/007786cad78c4de5dcd0e1f2c98cff47805acfcf/health/reducers/transcode.go#L33-L40)

Reply:
Gotcha thanks for the explanation.

I'm not sure if go-livepeer should be aware of the fact that it is a part of sharded infrastructure let alone a particular method of sharding based on a manifestID as would be the case here be exposing the notion of a manifestID based shard key. But, I don't have a better suggestion right now so okay with me since this the first pass!